File: PersianCalendar/src/main/java/com/byagowi/persiancalendar/ui/preferences/widgetnotification/FragmentWidgetNotification.java
package com.byagowi.persiancalendar.ui.preferences.widgetnotification;

import android.content.SharedPreferences;
import android.graphics.Color;
import android.os.Bundle;
import android.preference.PreferenceManager;

import com.byagowi.persiancalendar.R;

import java.util.Locale;

import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.FragmentActivity;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;

import static com.byagowi.persiancalendar.Constants.DEFAULT_SELECTED_WIDGET_TEXT_COLOR;
import static com.byagowi.persiancalendar.Constants.PREF_SELECTED_WIDGET_TEXT_COLOR;

// Don't use dagger in this class
public class FragmentWidgetNotification extends PreferenceFragmentCompat {
    @Override
    public void onCreatePreferences(Bundle bundle, String s) {
        addPreferencesFromResource(R.xml.preferences_widget_notification);
    }

    @Override
    public boolean onPreferenceTreeClick(Preference preference) {
        FragmentActivity activity = getActivity();
        if (activity == null) return false;

        SharedPreferences sharedPreferences = PreferenceManager.getDefaultSharedPreferences(activity);

        if (preference.getKey().equals(PREF_SELECTED_WIDGET_TEXT_COLOR)) {

            ColorPickerView colorPickerView = new ColorPickerView(activity);
            colorPickerView.setColorsToPick(
                    new int[]{0xFFFFFFFF, 0xFFE65100, 0xFF00796b, 0xFFFEF200, 0xFF202020});
            colorPickerView.setPickedColor(Color.parseColor(
                    sharedPreferences.getString(
                            PREF_SELECTED_WIDGET_TEXT_COLOR,
                            DEFAULT_SELECTED_WIDGET_TEXT_COLOR)));

            int padding = (int) (activity.getResources().getDisplayMetrics().density * 10);
            colorPickerView.setPadding(padding, padding, padding, padding);

            new AlertDialog.Builder(activity)
                    .setTitle(R.string.widget_text_color)
                    .setView(colorPickerView)
                    .setPositiveButton(R.string.accept, (dialogInterface, i) -> {
                        try {
                            SharedPreferences.Editor edit = sharedPreferences.edit();
                            edit.putString(PREF_SELECTED_WIDGET_TEXT_COLOR,
                                    String.format(Locale.ENGLISH, "#%06X",
                                            0xFFFFFF & colorPickerView.getPickerColor()));
                            edit.apply();
                        } catch (Exception e) {
                            e.printStackTrace();
                        }
                    }).setNegativeButton(R.string.cancel, null).show();
            return true;
        }
        return super.onPreferenceTreeClick(preference);
    }
}

File: PersianCalendar/src/main/java/com/byagowi/persiancalendar/ui/preferences/widgetnotification/FragmentWidgetNotification.kt
package com.byagowi.persiancalendar.ui.preferences.widgetnotification

import android.graphics.Color
import android.os.Bundle
import androidx.appcompat.app.AlertDialog
import androidx.core.content.edit
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import androidx.preference.PreferenceManager
import com.byagowi.persiancalendar.*
import java.util.*

// Don't use dagger in this class as it is used in WidgetConfigurationActivity also
class FragmentWidgetNotification : PreferenceFragmentCompat() {

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) =
        addPreferencesFromResource(R.xml.preferences_widget_notification)

    override fun onPreferenceTreeClick(preference: Preference?): Boolean {
        val activity = activity ?: return false

        val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(activity)

        if (preference?.key == PREF_SELECTED_WIDGET_TEXT_COLOR) {
            val colorPickerView = ColorPickerView(activity)
            colorPickerView.setColorsToPick(
                arrayOf(0xFFFFFFFF, 0xFFE65100, 0xFF00796b, 0xFFFEF200, 0xFF202020)
                    .map(Long::toInt).toIntArray()
            )
            colorPickerView.setPickedColor(
                Color.parseColor(
                    sharedPreferences.getString(
                        PREF_SELECTED_WIDGET_TEXT_COLOR,
                        DEFAULT_SELECTED_WIDGET_TEXT_COLOR
                    )
                )
            )
            colorPickerView.hideAlphaSeekbar()

            val padding = (activity.resources.displayMetrics.density * 10).toInt()
            colorPickerView.setPadding(padding, padding, padding, padding)

            AlertDialog.Builder(activity).apply {
                setTitle(R.string.widget_text_color)
                setView(colorPickerView)
                setPositiveButton(R.string.accept) { _, _ ->
                    try {
                        sharedPreferences.edit {
                            putString(
                                PREF_SELECTED_WIDGET_TEXT_COLOR,
                                String.format(
                                    Locale.ENGLISH, "#%06X",
                                    0xFFFFFF and colorPickerView.pickerColor
                                )
                            )
                        }
                    } catch (e: Exception) {
                        e.printStackTrace()
                    }
                }
                setNegativeButton(R.string.cancel, null)
            }.show()
            return true
        }

        if (preference?.key == PREF_SELECTED_WIDGET_BACKGROUND_COLOR) {
            val colorPickerView = ColorPickerView(activity)
            colorPickerView.setColorsToPick(
                arrayOf(0x00000000L, 0x50000000L, 0xFF000000L)
                    .map(Long::toInt).toIntArray()
            )
            colorPickerView.setPickedColor(
                Color.parseColor(
                    sharedPreferences.getString(
                        PREF_SELECTED_WIDGET_BACKGROUND_COLOR,
                        DEFAULT_SELECTED_WIDGET_BACKGROUND_COLOR
                    )
                )
            )

            val padding = (activity.resources.displayMetrics.density * 10).toInt()
            colorPickerView.setPadding(padding, padding, padding, padding)

            AlertDialog.Builder(activity).apply {
                setTitle(R.string.widget_background_color)
                setView(colorPickerView)
                setPositiveButton(R.string.accept) { _, _ ->
                    try {
                        sharedPreferences.edit {
                            putString(
                                PREF_SELECTED_WIDGET_BACKGROUND_COLOR,
                                String.format(
                                    Locale.ENGLISH, "#%08X",
                                    0xFFFFFFFF and colorPickerView.pickerColor.toLong()
                                )
                            )
                        }
                    } catch (e: Exception) {
                        e.printStackTrace()
                    }
                }
                setNegativeButton(R.string.cancel, null)
            }.show()
            return true
        }
        return super.onPreferenceTreeClick(preference)
    }
}
